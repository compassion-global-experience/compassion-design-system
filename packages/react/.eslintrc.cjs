module.exports = {
  extends: [
    'standard-jsx',
    'standard-react',
    'plugin:jsx-a11y/recommended',
    'plugin:react/recommended',
    'plugin:react/jsx-runtime',
    'plugin:react-hooks/recommended',
    'plugin:storybook/recommended',
  ],
  env: {
    browser: true,
    es6: true,
    jest: true,
    node: false,
  },
  rules: {
    'jsx-quotes': ['error', 'prefer-double'],
  },
};
