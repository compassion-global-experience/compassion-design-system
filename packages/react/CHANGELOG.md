# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.1.0 (2023-04-08)

### Features

- **react:** css usage and documentation ([11e317c](https://github.com/compassion-global-experience/compassion-design-system/commit/11e317cca4ecdaa385e86d3a5571d7dca554685d))
- **react:** re-export css package ([67549a5](https://github.com/compassion-global-experience/compassion-design-system/commit/67549a58de3de6e02b58c5bdb6f4605001aae783))
- **react:** switch between themes in storybook ([79119ba](https://github.com/compassion-global-experience/compassion-design-system/commit/79119ba48ed3410eaebe9a91ccd076b05c2580c1))
