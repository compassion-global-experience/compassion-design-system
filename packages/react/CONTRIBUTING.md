# Contributing to the `react-web` Package

The `react` package is structured with a component-per-directory convention
inside the `src` directory. Each component directory generally contains 3 items:

- A **TSX file** that describes the component‚Äôs structure and behavior
- A **.stories.ts file** that illustrates most of the ways the component can
  be configured
- An **index.ts** file that exports the TSX component‚Äôs contents

## Workflow

The package uses [Storybook](https://storybook.js.org/docs/react/get-started/introduction)
to facilitate component development and documentation. Storybook uses the files
mentioned above to render components in their various states, and serves as
documentation of those components by surfacing code comments, `propTypes`, and
`defaultProps` values.

To launch Storybook, make sure you‚Äôre in the [/packages/react](/packages/react)
directory and run the command `npm run storybook`. After the project is built,
Storybook will launch in your default browser. New components or modifications
will be displayed in real time! üéâ

The configurations outlined in each component‚Äôs .stories.ts file can be seen in
Storybook‚Äôs **Docs** tab.

## Commit Messages & Versioning

Every package in the compassion-gds monorepo adheres to
[semantic versioning](https://semver.org/). In order to simplify the versioning
and release process, please be sure your commit messages adhere to the
[‚ÄúConventional Commits‚Äù pattern](https://www.conventionalcommits.org/en/v1.0.0/#summary).
You may wish to install the [commitizen/cz-cli](https://github.com/commitizen/cz-cli)
to make this near-bulletproof.

## Making a Pull Request

Once your changes are ready, open a pull request against the `development`
branch of the project, and it will be merged in as quickly as possible!
