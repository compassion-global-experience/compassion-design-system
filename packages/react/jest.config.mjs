/*
 * For a detailed explanation regarding each configuration property, visit:
 * https://jestjs.io/docs/configuration
 */

export default {
  // A preset that is used as a base for Jest's configuration
  preset: 'ts-jest/presets/default-esm',

  // Automatically clear mock calls, instances, contexts and results before every test
  clearMocks: true,

  // Indicates whether the coverage information should be collected while executing the test
  collectCoverage: true,

  // The directory where Jest should output its coverage files
  coverageDirectory: "coverage",

  // Indicates which provider should be used to instrument code for coverage
  coverageProvider: "v8",

  // The test environment that will be used for testing
  testEnvironment: "jsdom",

  // A map from regular expressions to module names or to arrays of module names that allow to stub out resources with a single module
  moduleNameMapper: {
    ".(css|less|scss)$": "identity-obj-proxy",
  },

  // Whether to use watchman for file crawling
  watchman: true,
};
