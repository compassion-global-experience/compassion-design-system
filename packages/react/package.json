{
  "name": "@compassion-gds/react",
  "version": "0.2.0-alpha.2",
  "description": "",
  "type": "module",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "files": [
    "dist"
  ],
  "types": "dist/index.d.ts",
  "exports": {
    ".": {
      "browser": {
        "default": "./dist/esm/index.js"
      },
      "import": "./dist/esm/index.js",
      "require": "./dist/cjs/index.js"
    },
    "./css/base.css": "./dist/css/reset.css",
    "./css/light.css": "./dist/css/cds-light.css",
    "./css/dark.css": "./dist/css/cds-dark.css",
    "./css/*": "./dist/css/*"
  },
  "scripts": {
    "dev": "rollup -c -w",
    "copy-css": "bash ./scripts/copy-css.sh",
    "prepare": "rollup -c && npm run copy-css",
    "test": "jest",
    "storybook": "storybook dev -p 6007",
    "build:storybook": "storybook build --quiet",
    "percy": "cross-env PERCY_TOKEN=$PERCY_TOKEN_REACT percy storybook ./storybook-static",
    "percy-snapshots": "npm run build:storybook && npm run percy"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/compassion-global-experience/compassion-design-system.git",
    "directory": "packages/react"
  },
  "keywords": [
    "compassion",
    "compassion-gds",
    "gds",
    "react"
  ],
  "author": "Compassion XD",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/compassion-global-experience/compassion-design-system/issues"
  },
  "homepage": "https://github.com/compassion-global-experience/compassion-design-system#readme",
  "devDependencies": {
    "@babel/core": "^7.20.12",
    "@babel/preset-env": "^7.20.2",
    "@babel/preset-react": "^7.18.6",
    "@babel/preset-typescript": "^7.18.6",
    "@compassion-gds/core": "^0.1.0-alpha.0",
    "@rollup/plugin-commonjs": "^24.0.1",
    "@rollup/plugin-node-resolve": "^15.0.1",
    "@rollup/plugin-replace": "^5.0.2",
    "@rollup/plugin-typescript": "^11.1.0",
    "@storybook/addon-actions": "^7.0.2",
    "@storybook/addon-docs": "^7.0.2",
    "@storybook/addon-essentials": "^7.0.2",
    "@storybook/addon-interactions": "^7.0.2",
    "@storybook/addon-links": "^7.0.2",
    "@storybook/preset-react-webpack": "7.0.2",
    "@storybook/react": "7.0.2",
    "@storybook/react-webpack5": "7.0.2",
    "@storybook/testing-library": "^0.0.14-next.1",
    "@testing-library/react": "^13.4.0",
    "@types/jest": "^29.4.0",
    "@types/react": "^18.0.27",
    "babel-loader": "^8.3.0",
    "eslint": "^8.33.0",
    "eslint-config-standard-jsx": "^11.0.0",
    "eslint-config-standard-react": "^13.0.0",
    "eslint-plugin-jsx-a11y": "^6.7.1",
    "eslint-plugin-react": "^7.32.2",
    "eslint-plugin-react-hooks": "^4.6.0",
    "eslint-plugin-storybook": "^0.6.10",
    "jest": "^29.4.1",
    "jest-environment-jsdom": "^29.4.1",
    "phosphor-react": "^1.4.1",
    "react": "^18.2.0",
    "react-dom": "^18.2.0",
    "rollup": "^3.20.2",
    "rollup-plugin-dts": "^5.1.1",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-postcss": "^4.0.2",
    "storybook": "7.0.2",
    "ts-jest": "^29.0.5",
    "webpack": "^5.75.0"
  },
  "peerDependencies": {
    "react": "16.8.x || 17.x || 18.x",
    "react-dom": "16.8.x || 17.x || 18.x"
  }
}
