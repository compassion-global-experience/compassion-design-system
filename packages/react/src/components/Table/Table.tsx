import {
  ReactElement,
  CSSProperties,
  forwardRef,
  ForwardedRef,
  Ref,
} from 'react';
import '@compassion-gds/css/src/components/Table/table.module.css';

export interface Column<Row> {
  title: string;
  key: string;
  headerRender?: (col: Column<Row>) => ReactElement;
  cellRender?: (col: Column<Row>, data: Row) => ReactElement;
}

export interface TableProps<Row> {
  columns: Column<Row>[];
  rows: Row[];
  disabled?: boolean;
  stickyHeader?: boolean;
  containerStyle?: CSSProperties;
  tableStyle?: CSSProperties;
  className?: string;
}

function TableInner<Row>(
  props: TableProps<Row>,
  ref: ForwardedRef<HTMLTableElement>,
) {
  const {
    columns = [],
    rows = [],
    disabled = false,
    stickyHeader = false,
    containerStyle,
    tableStyle,
    className,
  } = props;

  const sticky = stickyHeader ? 'sticky-header' : '';
  const disabledTable = disabled ? 'disabled' : '';
  const classNames = ['table', sticky, disabledTable].filter(Boolean).join(' ');
  const containerClassNames = ['table-container', className]
    .filter(Boolean)
    .join(' ');

  return (
    <div className={containerClassNames} style={containerStyle}>
      <table ref={ref} className={classNames} style={tableStyle}>
        <thead className="table-head">
          <tr className="table-row">
            {columns.map((col) =>
              col.headerRender ? (
                col.headerRender(col)
              ) : (
                <th key={col.key} className="table-cell">
                  {col.title}
                </th>
              ),
            )}
          </tr>
        </thead>
        <tbody className="table-body">
          {rows.map((row, rowIndex) => (
            <tr key={rowIndex} className="table-row">
              {columns.map((col) =>
                col.cellRender ? (
                  col.cellRender(col, row)
                ) : (
                  <td className="table-cell" key={`${rowIndex}-${col.key}`}>
                    {row[col.key]}
                  </td>
                ),
              )}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}

const TableWithRef = forwardRef(TableInner);

type TableWithRefProps<T> = TableProps<T> & {
  mRef?: Ref<HTMLTableElement>;
};

function Table<RowData>({ mRef, ...props }: TableWithRefProps<RowData>) {
  return <TableWithRef ref={mRef} {...props} />;
}

export default Table;
