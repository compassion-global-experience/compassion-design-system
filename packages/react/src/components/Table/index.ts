export { default, Column, TableWithRefProps } from './Table';
