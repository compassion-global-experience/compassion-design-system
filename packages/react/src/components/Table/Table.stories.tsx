import React from 'react';
import { Meta } from '@storybook/react';
import Table from './Table';

const meta: Meta<typeof Table> = {
  title: 'Components/Table',
  component: Table,
};

export default meta;

export const Default = {
  args: {
    containerStyle: { height: 150, width: 400 },
    tableStyle: { width: 500 },
    columns: [
      { value: 'Heading 1' },
      { value: 'Heading 2', link: 'https://example.com' },
      { value: 'Heading 3' },
    ],
    rows: [
      {
        data: [
          { value: 'Cell 1' },
          { value: 'Cell 2' },
          { value: 'Cell 3' },
        ],
        disabled: false,
      },
      {
        data: [
          { value: 'Link 1', link: 'https://example.com' },
          { value: 'Link 2', link: 'https://example.com' },
          { value: 'Link 3', link: 'https://example.com' },
        ],
        disabled: false,
      },
      {
        data: [
          { value: 'Cell 1' },
          { value: 'Cell 2' },
          { value: 'Cell 3' },
        ],
        disabled: false,
      },
    ],
    stickyHeader: false,
  },
};
