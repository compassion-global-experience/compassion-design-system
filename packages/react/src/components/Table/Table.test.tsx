import { render } from '@testing-library/react';
import Table from './Table';

describe('Table', () => {
  it('should render 3 columns and 3 rows', () => {
    const { container } = render(<Table {...TABLE_DEFAULT} />);
    expect(container).toMatchSnapshot();
  });

  it('should render disabled version', () => {
    const { container } = render(<Table {...TABLE_DEFAULT} disabled />);
    expect(container).toMatchSnapshot();
  });

  it('should render sticky header version', () => {
    const { container } = render(<Table {...TABLE_DEFAULT} stickyHeader />);
    expect(container).toMatchSnapshot();
  });
});

const TABLE_DEFAULT = {
  columns: [
    { title: 'Heading 1', key: 'column1' },
    { title: 'Heading 2', key: 'column2' },
    { title: 'Heading 3', key: 'column3' },
  ],
  rows: [
    { column1: 'Cell 1', column2: 'Cell 2', column3: 'Cell 3' },
    { column1: 'Cell 1', column2: 'Cell 2', column3: 'Cell 3' },
    { column1: 'Cell 1', column2: 'Cell 2', column3: 'Cell 3' },
  ],
  stickyHeader: false,
};
