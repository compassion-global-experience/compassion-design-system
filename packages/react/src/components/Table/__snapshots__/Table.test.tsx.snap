// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render 3 columns and 3 rows 1`] = `
<div>
  <div
    class="cds-table__container"
  >
    <table
      class="cds-table"
    >
      <thead
        class="cds-table__head"
      >
        <tr
          class="cds-table__row"
        >
          <th
            class="cds-table__cell"
          >
            Heading 1
          </th>
          <th
            class="cds-table__cell"
          >
            Heading 2
          </th>
          <th
            class="cds-table__cell"
          >
            Heading 3
          </th>
        </tr>
      </thead>
      <tbody
        class="cds-table__body"
      >
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render disabled version 1`] = `
<div>
  <div
    class="cds-table__container"
  >
    <table
      class="cds-table cds-table--disabled"
    >
      <thead
        class="cds-table__head"
      >
        <tr
          class="cds-table__row"
        >
          <th
            class="cds-table__cell"
          >
            Heading 1
          </th>
          <th
            class="cds-table__cell"
          >
            Heading 2
          </th>
          <th
            class="cds-table__cell"
          >
            Heading 3
          </th>
        </tr>
      </thead>
      <tbody
        class="cds-table__body"
      >
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render sticky header version 1`] = `
<div>
  <div
    class="cds-table__container"
  >
    <table
      class="cds-table cds-table__sticky-header"
    >
      <thead
        class="cds-table__head"
      >
        <tr
          class="cds-table__row"
        >
          <th
            class="cds-table__cell"
          >
            Heading 1
          </th>
          <th
            class="cds-table__cell"
          >
            Heading 2
          </th>
          <th
            class="cds-table__cell"
          >
            Heading 3
          </th>
        </tr>
      </thead>
      <tbody
        class="cds-table__body"
      >
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
        <tr
          class="cds-table__row"
        >
          <td
            class="cds-table__cell"
          >
            Cell 1
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 2
          </td>
          <td
            class="cds-table__cell"
          >
            Cell 3
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
